driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.async;
+
+import com.mongodb.lang.Nullable;
+
+import java.util.function.Predicate;
+
+
+/**
+ * See tests for usage (AsyncFunctionsTest).
+ * <p>
+ * This class is not part of the public API and may be removed or changed at any time
+ */
+@FunctionalInterface
+public interface AsyncSupplier<T> extends AsyncFunction<Void, T> {
+    /**
+     * This should not be called externally to this API. It should be
+     * implemented as a lambda. To "finish" an async chain, use one of
+     * the "finish" methods.
+     *
+     * @see #finish(SingleResultCallback)
+     */
+    void unsafeFinish(SingleResultCallback<T> callback);
+
+    /**
+     * This method must only be used when this AsyncSupplier corresponds
+     * to a {@link java.util.function.Supplier} (and is therefore being
+     * used within an async chain method lambda).
+     * @param callback the callback
+     */
+    default void getAsync(final SingleResultCallback<T> callback) {
+        unsafeFinish(callback);
+    }
+
+    @Override
+    default void unsafeFinish(@Nullable final Void value, final SingleResultCallback<T> callback) {
+        unsafeFinish(callback);
+    }
+
+    /**
+     * Must be invoked at end of async chain.
+     * @param callback the callback provided by the method the chain is used in
+     */
+    default void finish(final SingleResultCallback<T> callback) {
+        final boolean[] callbackInvoked = {false};
+        try {
+            this.unsafeFinish((v, e) -> {
+                callbackInvoked[0] = true;
+                callback.onResult(v, e);
+            });
+        } catch (Throwable t) {
+            if (callbackInvoked[0]) {
+                throw t;
+            } else {
+                callback.onResult(null, t);
+            }
+        }
+    }
```

Comment:
I comment [above](https://github.com/mongodb/mongo-java-driver/pull/1131), where I think you refer to your comment here. I had said:

> Unaffected async code SHOULD be preserved, including for example null checks, which throw exceptions, so chained lambdas MAY throw exceptions, which are expected to be handled in exactly one place (the finish method, when it is invoked at the end of the chain).

To put it another way, if we do (1), "behavesSame" tests in AsyncFunctionsTest will start to fail, and if we do (2), `testInvalid` fails.

It seems to me that throwing exceptions is a "bug" in async code only because those exceptions can be lost. But, if the API handles them, there is no issue, and no bug. The API can either handle and allow them, or treat them as bugs (and throw assertion errors). Both are acceptable approaches. However, treating them as bugs would mean that the user will need to create boilerplate or special constructs to asynchronously handle exceptions, which is inconvenient and error-prone, so the first alternative should be preferred.

Reply:
> Unaffected async code SHOULD be preserved, including for example null checks, which throw exceptions, so chained lambdas MAY throw exceptions

I can't find a justification for this statement on my own: why should the behavior of incorrect code be preserved? Internal null-checks should produce `AssertionError` rather than `NullPointerException` (the latter being an `Exception` is an utter nonsense: NPE is always a software bug, not a recoverable exception), but even if they produce `NullPointerException`, all that means is "the driver's code is broken".

> It seems to me that throwing exceptions is a "bug" in async code only because those exceptions can be lost.

If this were true, then we could just always throw all exceptions normally in async code, and make sure we log then at some point, e.g., by setting a logging `UncaughtExceptionHandler` for threads.

The reason the contracts of both `AsyncFunction.unsafeFinish` and `SingleResultCallback.onResult` must not allow these methods to complete abruptly with an `Exception` is that it violates the very basic code behavior, making programming virtually impossible: `AsyncFunction.unsafeFinish`/`SingleResultCallback.onResult` completing abruptly is equivalent to an exception being thrown out of the blue in synchronous code, and at worst not executing any instructions that follow in program order, including `final` blocks and returning from a method call<sup>1</sup>.

> To put it another way, if we do (1), "behavesSame" tests in AsyncFunctionsTest will start to fail, and if we do (2), testInvalid fails.

Executions where `AsyncFunction.unsafeFinish`/`SingleResultCallback.onResult` comple abruptly can't generally be compared with synchronous code because the behavior I described above is simply impossible in synchronous Java code. I.e., the tests comparing behavior of sync and async code are not appropriate here.

> if the API handles them, there is no issue, and no bug

The way it is currently done is not a panacea:

1. If `AsyncSupplier.finish` can't pass `t` to `callback.onResult` because `callback.onResult` has already been called, then `t` is neither handled nor can it be.
2. If `AsyncSupplier.finish` passes `t` to `callback.onResult`, then it is very much possible that this `t` was actually meant for a different callback, i.e., this may be equivalent to an exception being thrown in synchronous code out of the blue.

Neither of the above situations is fine, they are both software bugs. #1 is obvious, #2 may be less obvious, so I am providing an example:

1. Consider `DefaultConnectionPool.getAsync` being called as
```java
beginAsync()
    .thenSupply(c -> pool.getAsync(operationContext, c))
    .finish(callbackForGetAsync)
```
2. Imagine that here https://github.com/mongodb/mongo-java-driver/blob/4ab6048bc5f5d6593ee69814092a8b59f6c28391/driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java#L220-L225 the code incorrectly throwing `e` instead of calling `eventSendingCallback.onResult(null, e)`.
3. `finish` catches `e` and passes it to `callbackForGetAsync`, but `e` was meant to be handled by `eventSendingCallback.onResult`, which has never been called, i.e., `checkOutFailed` has not been called.
5. As you can see, `finish` did not make things right despite passing `e` to `callbackForGetAsync`.

---

<sup>1</sup> It's also possible for `AsyncFunction.unsafeFinish`/`SingleResultCallback.onResult` to complete abruptly after calling the callback (a wrapped callback in case of `onResult` if there is one), in which case coming up with a synchronous analogy appears to be impossible.

Reply:
Discussed a bit, leaving as is.

The two cases you mention are programming errors, present independently of the API. There is an opportunity to wrap exceptions in assertion errors, to better-highlight driver bugs. This would imply that any "expected" ("good") exceptions must be accurately discovered and explicitly pushed into callbacks - that is, async code would need to be "marked up" with the locations of good exceptions, and all others treated as "bad". This would complicate the async code, and complication increases the likelihood of bugs (introducing such boilerplate makes it harder to compare sync and async, which is the primary goal). It is also imperfect: arguably, indiscriminately turning what would otherwise be an "expected" exception into an assertion error is worse than indiscriminately propagating without wrapping.

The alternative taken here is to effectively wrap async code sections in a "try catch", and propagate all exceptions into the callback "as they are". This implies that it is no longer a coding error to throw "expected" exceptions from async code that is used in this API.

Reply:
> This would imply that any "expected" ("good") exceptions must be accurately discovered and explicitly pushed into callbacks

Any exceptions thrown by a callback-based method or `SingleResultCallback.onResult` is not expected and signifies a bug. 

>  This implies that it is no longer a coding error to throw "expected" exceptions from async code that is used in this API.

It still is, and as far as I can see, nothing can be done to change that.